Basic range download handling for Force download.

peterfabian opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

This change adds support for HTTP_RANGE requests (used by iOS devices) to `Force download` method.
One thing to note is that range request downloads are not logged (so that they are not counted as multiple downloads). The most reasonable option in my view would be to add extra column(s) to table `wp_wc_download_log`, e.g. columns `start` and `length` and then exclude range request log entries when counting number of downloads. 
Not sure how feasible db changes are mid-release-cycle, so I went for easier option for now--ignoring ranged downloads from the log. 
As the initial request from an iOS device is a normal (non-range) request, the download should be logged once anyway.
 
Closes #19640 .

### How to test the changes in this Pull Request:

1. Create a downloadable virtual product which contains larger audio or video file, e.g. an mp3 file.
2. Buy the product, go to My account > Downloads.
3. Try to download the mp3 file on iPhone/iPad.
4. Audio only plays first ~2 minutes, then starts from the beginning again, it does not indicate the length and does not allow seeking.
5. Apply this branch.
6. See the audio now shows correct length and allows seeking.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable? No, it's a bit tricky as headers interfere with phpunit's output and e2e tests don't run on iOS device.
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Added support for HTTP_RANGE requests (used by iOS devices) to `Force download` method.

peterfabian committed:
Basic range download handling for Force download.

tiagonoronha commented:
Tested and can confirm it's no longer showing up as a "Live Broadcast".

peterfabian committed:
Split handle_range_request function, code cleanup.

peterfabian committed:
Backwards compat fixes for range download.

peterfabian committed:
Sanitize HTTP_RANGE input.

peterfabian committed:
Updated documentation according to WP CS.

peterfabian committed:
Added test for range download logging.

peterfabian committed:
Fixed PHPCS problems in test file.

peterfabian committed:
Fixed bugs introduced by refactoring.

mikejolley commented on includes/class-wc-customer-download.php:
```diff
@@ -290,15 +290,21 @@ public function set_download_count( $value ) {
+	 * @param bool   $range_download  Whether the download is a partial/range download.
 	 */
-	public function track_download( $user_id = null, $user_ip_address = null ) {
+	public function track_download( $user_id = null, $user_ip_address = null, $range_download = false ) {
```
If track_download does not run for `range_download` (which I agree is the correct handling) why is this passed through?

Could this check not go in the download handler?

peterfabian replied:
Good point, I initially thought that we might be updating the ranges in the `wp_wc_download_log`, but since that's not the way it ended up, it can be removed altogether.

peterfabian committed:
Moved check for download log to WC_Download_Handler.

peterfabian committed:
Removed test to reflect changes in track download method.

peterfabian committed:
Skip download remainig check if doing ranged request.

This allows to finish listening/waching the file on iOS device if the remaining download count goes from 1 to 0.

mikejolley reviewed (approved):
Works for me now 👍 

@claudiulodro you should check too as this is larger than other PRs.

claudiulodro commented:
This looks good to me also. Works well in my testing and I don't notice anything wrong in the code.
